From 26084b61e3d5cb814f2e2e7ae97435716d51d673 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 31 Jan 2024 11:40:28 +0800
Subject: [PATCH 41/87] mtk: mac80211: remove link != 0 warn on in
 rate_control_rate_update for mlo channel switch

Remove link warning for mlo channel switch

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 net/mac80211/rate.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/rate.c b/net/mac80211/rate.c
index f974ad58..5558c185 100644
--- a/net/mac80211/rate.c
+++ b/net/mac80211/rate.c
@@ -1,4 +1,3 @@
-// SPDX-License-Identifier: GPL-2.0-only
 /*
  * Copyright 2002-2005, Instant802 Networks, Inc.
  * Copyright 2005-2006, Devicescape Software, Inc.
@@ -121,11 +120,17 @@ void rate_control_rate_update(struct ieee80211_local *local,
 	struct ieee80211_sta *ista = &sta->sta;
 	void *priv_sta = sta->rate_ctrl_priv;
 	struct ieee80211_chanctx_conf *chanctx_conf;
+	struct ieee80211_link_data *link;
 
 	if (ref && ref->ops->rate_update) {
 		rcu_read_lock();
 
-		chanctx_conf = rcu_dereference(sta->sdata->vif.bss_conf.chanctx_conf);
+		link = rcu_dereference(sta->sdata->link[link->link_id]);
+		if (!link) {
+			rcu_read_unlock();
+			return;
+		}
+		chanctx_conf = rcu_dereference(link->conf->chanctx_conf);
 		if (WARN_ON(!chanctx_conf)) {
 			rcu_read_unlock();
 			return;
-- 
2.45.2

